src/engraving/dom/dynamic.cpp

```diff
@@ -113,6 +141,127 @@ static const ElementStyle dynamicsStyle {
     { Sid::centerOnNotehead, Pid::CENTER_ON_NOTEHEAD },
 };
 
+//---------------------------------------------------------
+//   DynamicFragment
+//---------------------------------------------------------
+
+DynamicFragment::DynamicFragment(CharFormat charFormat, String dynamicText, bool dynamicDisplay)
+{
+    format = charFormat;
+    m_dynamicText = dynamicText;
+    calculatePlainText(getDynamics(dynamicText));
+
+    if (dynamicDisplay) {
+        setToDynamicText();
+    } else {
+        setToExpressionText();
+    }
+}
+
+DynamicFragment::DynamicFragment(const DynamicFragment& dynamicFragment)
+    : TextFragment(dynamicFragment)
+{
+    m_dynamicTextMode = dynamicFragment.m_dynamicTextMode;
+    m_dynamicText = dynamicFragment.m_dynamicText;
+    m_plainText = dynamicFragment.m_plainText;
+}
+
+DynamicFragment& DynamicFragment::operator =(const DynamicFragment& dynamicFragment)
+{
+    TextFragment::operator =(dynamicFragment);
+    m_dynamicText = dynamicFragment.m_dynamicText;
+    m_plainText = dynamicFragment.m_plainText;
+
+    return *this;
+}
+
+void DynamicFragment::setText(String newText)
+{
+    std::vector<DynamicType> dynamics = getDynamics(newText);
+    calculateDynamicText(dynamics);
+    calculatePlainText(dynamics);
+    if (m_dynamicTextMode) {
+        setToDynamicText();
+    } else {
+        setToExpressionText();
+    }
+}
+
+std::vector<DynamicType> DynamicFragment::getDynamics(String text)
+{
+    static std::shared_ptr<IEngravingFontsProvider> provider = muse::modularity::globalIoc()->resolve<IEngravingFontsProvider>("engraving");
+
+    std::vector<DynamicType> dynamics;
+    for (char32_t i = 0; i < text.size(); i++) {
+        SymId symId = provider->fallbackFont()->fromCode(text[i]);
+        DynamicType dt = TConv::dynamicType(symId);
+        if (dt == DynamicType::OTHER) {
+            return dynamics;
+        }
+        dynamics.push_back(dt);
+    }
+
+    return dynamics;
+}
+
+void DynamicFragment::setToDynamicText()
+{
+    format.setFontFamily(u"ScoreText");
+    TextFragment::setText(m_dynamicText);
+    m_dynamicTextMode = true;
+}
+
+void DynamicFragment::setToExpressionText()
+{
+    TextFragment::setText(m_plainText);
+    m_dynamicTextMode = false;
+}
+
+void DynamicFragment::setDynamicTextMode(bool dynamicTextMode)
+{
+    if (dynamicTextMode) {
+        setToDynamicText();
+    } else {
+        setToExpressionText();
+    }
+}
```

Comment:
This function simply calls the two functions above it. I think it would be better to move their functionality into this one and rename it as follows:

```C++
void DynamicFragment::setDisplayAsExpressionText(bool asExpression)
{
    m_displayAsExpressionText = asExpression;
    if (asExpression) {
        TextFragment::setText(m_plainText);
    } else {
        format.setFontFamily(u"ScoreText");
        TextFragment::setText(m_dynamicText);
    }
}
```

Notice I also renamed `m_dynamicTextMode` to `m_displayAsExpressionText`.